nixos/modules/services/matrix/synapse.nix

```diff
@@ -50,6 +45,27 @@ let
             "${bindAddress}"
         }:${builtins.toString listener.port}/"
     '';
+
+  wantsModule = module:
+    cfg.settings ? modules && lib.any (m: m.module == module) cfg.settings.modules;
+
+  wantedExtras = cfg.extras
+    ++ lib.optional (cfg.settings ? oidc_providers) "oidc"
+    ++ lib.optional (cfg.settings ? jwt_config) "jwt"
+    ++ lib.optional (cfg.settings ? saml2_config) "saml2"
+    ++ lib.optional (cfg.settings ? opentracing) "opentracing"
+    ++ lib.optional (cfg.settings ? redis) "redis"
+    ++ lib.optional (cfg.settings ? sentry) "sentry"
+    ++ lib.optional (cfg.settings ? user_directory) "user-search"
+    ++ lib.optional (cfg.settings.url_preview_enabled) "url-preview"
+    ++ lib.optional (cfg.settings.database.name == "psycopg2") "postgres"
+    ++ lib.optional (wantsModule "ldap_auth_provider.LdapAuthProviderModule") "matrix-synapse-ldap3";
```

Comment:
This would break when configuring LDAP auht like this, correct?

```
      password_providers = [
        { module = "ldap_auth_provider.LdapAuthProvider";
          config = {};
        }
      ]
```

Reply:
I wasn't aware you could configure it like that. From https://github.com/matrix-org/matrix-synapse-ldap3/ it looks like `modules` is they way to go.

And looks like the described way is the legacy way to setting this up. https://github.com/matrix-org/synapse/blob/a5fb382a29991c8eafcb8c54cdd8c7aab260c237/synapse/handlers/auth.py#L1858

Reply:
Heh good to know :) 
But anyways, just realized that that's actually not relevant because right now you have to add the LDAP module manually via `plugins` anyways, so nothing to be broken here ;-)

Reply:
I think we shouldn't expose the `matrix-synapse-ldap3` as an extra. 

1. It blurs the scope with regard to plugins and is really an outlier
2. We ideally want to run all tests, which means passing all extras into the test environment. With `matrix-synapse-ldap3` there is the problem, that it causes an infinite recursion with synapse itself.